cache/cache.go

```diff
@@ -181,86 +168,41 @@ func (c *cache[T]) set(key string, object T) {
 	c.items = append(c.items, &item)
 }
 
-// Get an item from the cache. Returns the item or nil, and a bool indicating
-// whether the key was found.
-func (c *Cache[T]) Get(object T) (item T, exists bool, err error) {
-	var res T
-	lvs := []string{}
-	if c.labelsFunc != nil {
-		lvs, err = c.labelsFunc(object, len(c.metrics.getExtraLabels()))
-		if err != nil {
-			recordRequest(c.metrics, StatusFailure)
-			return res, false, &CacheError{Reason: ErrInvalidLabels, Err: err}
-		}
-	}
-	key, err := c.keyFunc(object)
-	if err != nil {
-		recordRequest(c.metrics, StatusFailure)
-		return res, false, &CacheError{Reason: ErrInvalidKey, Err: err}
-	}
-	item, found, err := c.get(key)
-	if err != nil {
-		return res, false, err
-	}
-	if !found {
-		recordEvent(c.metrics, CacheEventTypeMiss, lvs...)
-		return res, false, nil
-	}
-	recordEvent(c.metrics, CacheEventTypeHit, lvs...)
-	return item, true, nil
-}
-
-// GetByKey returns the object for the given key.
-func (c *Cache[T]) GetByKey(key string) (T, bool, error) {
-	var res T
-	index, found, err := c.get(key)
-	if err != nil {
-		return res, false, err
-	}
-	if !found {
-		recordEvent(c.metrics, CacheEventTypeMiss)
-		return res, false, nil
-	}
-
-	recordEvent(c.metrics, CacheEventTypeHit)
-	return index, true, nil
-}
-
-func (c *cache[T]) get(key string) (T, bool, error) {
-	var res T
+// Get returns a pointer to an item in the cache for the given key. If no item
+// is found, it's a nil pointer.
+// The caller can record cache hit or miss based on the result with
+// Cache.RecordCacheEvent().
+func (c *Cache[T]) Get(key string) (*T, error) {
 	c.mu.RLock()
 	if c.closed {
 		c.mu.RUnlock()
 		recordRequest(c.metrics, StatusFailure)
-		return res, false, ErrCacheClosed
+		return nil, ErrCacheClosed
 	}
 	item, found := c.index[key]
 	if !found {
 		c.mu.RUnlock()
 		recordRequest(c.metrics, StatusSuccess)
-		return res, false, nil
+		return nil, nil
```

Comment:
In the previous implementation the bool return value (exists) indicated if the element is not present in the cache. With the current implementation, the caller cannot deduce a cache miss scenario since no error is returned when the cache element is not found. This leads to a segmentation fault in the caller while dereferencing the item. This can be fixed by returning ErrNotFound here if the item is not found in the cache. The caller can then check for ErrNotFound to figure out if this is a cache miss.

Reply:
Ok, my vote is to return an error for not found and not return a pointer as it feels more intuitive to use and keeps the implementation in sync with the expiration methods.